### PR TITLE
Increase the precision of the exportable MIR check

### PR DESCRIPTION
**Pull request by saethlin**
Before this PR, MIR inlining only inlines across crates or within a crate if the callee is a constructor, `#[inline]`, or generic. This misses some inlining opportunities within a crate; trivial helper functions are not always given `#[inline]` (and maybe that's a good thing).

The reason we were so conservative before is that the MIR inliner needs to not drag MIR which mentions items that we did not generate MIR for, or items that we have given internal linkage, into bodies that we must generate MIR for. If we get this heuristic wrong, we see ICEs (for promoted consts without MIR) or linker errors (for private functions and statics now mentioned in exported MIR). So this PR adds more fine-grained analysis to detect callee bodies that contain MIR we must not export.

Perhaps even better, this PR removes any prohibition on inlining not-exported bodies into other not-exported bodies.

### CODE REVIEW

**Comment by est31** [comment]
@craterbot run mode=build-only start=5683791ebb7bbc440366635c322ad82badc554ca end=f2c6735504f9c1c0baa07ad5932d994d4c8be6db

as per [request on zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/crater.20run).

**Comment by Mark-Simulacrum** [comment]
@craterbot abort

**Comment by Mark-Simulacrum** [comment]
@craterbot run mode=build-only start=master#5683791ebb7bbc440366635c322ad82badc554ca end=try#f2c6735504f9c1c0baa07ad5932d994d4c8be6db

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -558,6 +558,8 @@ pub struct GlobalCtxt<'tcx> {
 
     /// Stores memory for globals (statics/consts).
     pub(crate) alloc_map: Lock<interpret::AllocMap<'tcx>>,
+
+    pub inlined_internal_defs: Lock<FxHashSet<LocalDefId>>,
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:562]: This field will need to be removed before merging, for incr. comp. support.

Comment (saethlin) [reply compiler/rustc_middle/src/ty/context.rs:562]: Why?

The most recent commit in here is a hack to make crater exercise this PR without having to do a `--release` run.

**Comment by saethlin** [comment]
What is the alternative implementation? I don't see anything suggested in that issue.

**Comment by saethlin** [comment]
I'm confused. This looks like it is based on `instance_mir`, which is implemented by calling `optimized_mir`: https://github.com/rust-lang/rust/blob/939786223f2d36b1af62e27e9a7a54bae6e30f3f/compiler/rustc_monomorphize/src/collector.rs#L1367-L1376

Am I looking at the wrong collector?

File: compiler/rustc_middle/src/ty/context.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_middle/src/ty/context.rs:562]: In incremental mode, optimized MIR may get loaded from the on-disk cache, and the insertions into this map may not be re-executed.

**Comment by saethlin** [comment]
I think all the regressions are because this PR enables MIR inlining in all debug builds. All the regressions already produce the same error or ICE with `cargo +nightly test --release`.

**Comment by cjgillot** [comment]
IIUC, the goal of this PR is to encore promoted_mir, in case we inline a body which references a promoted constant. Instead of relying on a global inlined_internal_defs, can this be performed by walking `required_consts` in the metadata encoder?

**Comment by saethlin** [comment]
@tmiasko I think the problem you're trying to point me at is with the implementation of `reachable_set` here: https://github.com/rust-lang/rust/blob/f20afcc455bbcc5c0f7679450fb35fd0c9668880/compiler/rustc_passes/src/reachable.rs

Because `MirUsedCollector` calls `is_reachable_non_generic`:
https://github.com/rust-lang/rust/blob/f20afcc455bbcc5c0f7679450fb35fd0c9668880/compiler/rustc_monomorphize/src/collector.rs#L949
which is in turn based on `reachable_set`:
https://github.com/rust-lang/rust/blob/f20afcc455bbcc5c0f7679450fb35fd0c9668880/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L60-L61

So I'm trying to reimplement `reachable_set` based on optimized MIR, and while I'm at it, I'm trying to remove this thing that looks to me like a hack (based on the fact that it has a FIXME) but might also not be (based on the fact that as far as I can tell, the suggested fix doesn't work):
https://github.com/rust-lang/rust/blob/f20afcc455bbcc5c0f7679450fb35fd0c9668880/compiler/rustc_passes/src/reachable.rs#L396-L409

**Comment by JohnCSimon** [comment]
@saethlin 

ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!

**Comment by saethlin** [comment]
Yes.

I'm closing this PR because the work that's described in these PR comments is actually in a different branch of mine. If I make progress on this whole concept, it will be by landing that branch first, then re-pursuing this idea.

**Pull request closed without merging**